### DISCUSSION TITLE
How do I query a nested collection?

### DISCUSSION THREAD

**Discussion by pndalal** [GodotNuts/GodotFirebase#236] [Q&A] [answered]
I wasn't able to find an example of this anywhere. If I try:

`query.from("questions/some-id/my-nested-collection")`

I get the following error:

```
[Firebase Error] >> Collection id "questions/dmGlFrcaF7CHVxSzUjy6/user_responses" is invalid because it contains "/".
{code:400, message:Collection id "questions/dmGlFrcaF7CHVxSzUjy6/user_responses" is invalid because it contains "/"., status:INVALID_ARGUMENT}
```

**Comment by fenix-hub** [comment, marked as answer]
Hi @pndalal ,
Have you tried replacing `\` with `/` ?

**Reply by pndalal** [reply]
Ah my bad, I wrote the question incorrectly. I've edited it to include the correct slash. Below is the error I get.

```
[Firebase Error] >> Collection id "questions/dmGlFrcaF7CHVxSzUjy6/user_responses" is invalid because it contains "/".
{code:400, message:Collection id "questions/dmGlFrcaF7CHVxSzUjy6/user_responses" is invalid because it contains "/"., status:INVALID_ARGUMENT}
```

I tried with the opposite slash for fun, and while it accepts this as a parameter, I'm fairly certain it isn't the way to query a nested collection. I'm looking for a concrete example that works.

**Reply by fenix-hub** [reply]
Pretty sure this needs some investigation, I'll make an issue out of this so you can get along.

**Reply by WolfgangSenff** [reply]
You can't query the sub-collection, as far as I know, without first getting a reference to the sub-collection. You can use "/" in paths to sub-collections (you'd just navigate all the way down the chain like you have above), and then you'd query at that level. Here's how I did it:

 `var collection = Firebase.Firestore.collection(series + "/" + firestore_id + "/" + collection_name)
 ` 

Note those are all variables that I pass into the function that creates the collection, and I believe you should be able to query this then.

**Reply by fenix-hub** [reply]
> You can't query the sub-collection, as far as I know, without first getting a reference to the sub-collection. You can use "/" in paths to sub-collections (you'd just navigate all the way down the chain like you have above), and then you'd query at that level. Here's how I did it:
> 
> `var collection = Firebase.Firestore.collection(series + "/" + firestore_id + "/" + collection_name) `
> 
> Note those are all variables that I pass into the function that creates the collection, and I believe you should be able to query this then.

I'm pretty sure that's the proper way for now. 
The error 
```
[Firebase Error] >> Collection id "questions/dmGlFrcaF7CHVxSzUjy6/user_responses" is invalid because it contains "/".
{code:400, message:Collection id "questions/dmGlFrcaF7CHVxSzUjy6/user_responses" is invalid because it contains "/"., status:INVALID_ARGUMENT}
```
comes directly from Firebase, not our plugin, so it's Firebase itself that doesn't accept this kind of navigation in queries.

**Reply by pndalal** [reply]
Thank you for your answers, but I'm not following the logic here. 

Even if I get a reference to the sub-collection, all I can do is add, update, delete, or get specific documents by using methods on that reference. I still do not see how I can _query_ the sub-collection using that reference (using "where" and "limit" and all that), because the reference contains no querying methods. If I attempt to use a query independently, after obtaining a reference, I get the same error as usual.

@fenix-hub This is indeed a Firebase error, but allow me to clarify my concern. I believe Firebase does allow a sub-collection to be queried, one way or the other. Currently, it may be that the plugin is not built to support that functionality, or perhaps the REST API makes this functionality more complicated. But if Firebase didn't allow a sub-collection to be queried, then sub-collections would not be very useful, I think.

Either way, I am no longer blocked, per say. For now, I've decided to restructure my data to not use sub-collections, because there doesn't seem to be an immediate solution. I do believe that querying a sub-collection is standard functionality that the plugin should support at some point. If I get the time later, I can investigate this further myself.
